### PR TITLE
[runtime env] plugin refactor[2/n]: support json schema validation

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?

The second PR of runtime env plugin refactor. The prototype is https://github.com/ray-project/ray/pull/25387.

- Use json schema to validate the runtime env.
- Add plugin_schema_manager.py which is used to manage the schemas in workers. It will load all of the schemas lazily when the first time use of runtime_env.
- Add default schemas: pip_schema.json and working_dir_schema.json.
- Set the env var `RAY_RUNTIME_ENV_PLUGINS_SCHEMAS ` to append external schemas.

Future works:
- Add other schema files, such as conda, py_modules, container...
- Remove the existing validation logic in `python/ray/_private/runtime_env/validation.py`
- Support validation in ray-client/server mode. (There are no schemas files of plugins in ray client side.)

### CODE REVIEW

File: python/ray/_private/runtime_env/plugin_schema_manager.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        if schema_paths:
+            schema_json_files = list()
+            for path in schema_paths.split(","):
+                if path.endswith("schema.json"):
+                    schema_json_files.append(path)
+                elif os.path.isdir(path):
+                    for root, _, files in os.walk(path):
+                        for f in files:
+                            if f.endswith("schema.json"):
+                                schema_json_files.append(os.path.join(root, f))
```
[the remark this answers is not in the document]
Comment (architkulkarni) [reply python/ray/_private/runtime_env/plugin_schema_manager.py:62]: On the other hand, maybe printing a warning is too spammy and isn't the best practice.  If I remember from the examples of `PATH` and `PYTHONPATH`, nonexistent directories are silently ignored.

**Review by architkulkarni — COMMENTED** [review]
Overall looks great!  Just a few comments.

File: python/ray/_private/runtime_env/plugin_schema_manager.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                )
+                continue
+            cls.schemas[schema["title"]] = schema
+
+    @classmethod
+    def _load_default_schemas(cls):
+        schema_json_files = list()
+        for root, _, files in os.walk(cls.default_schema_path):
+            for f in files:
+                if f.endswith("schema.json"):
```
[the remark this answers is not in the document]
Comment (SongGuyang) [reply python/ray/_private/runtime_env/plugin_schema_manager.py:41]: My concern is that if some user put many kind of json config files into one directory, we will load a lot unrelated config files. But I don't strongly think of that. Maybe I can use ".json" first in current PR and optimize it in future if needed.

File: python/ray/_private/runtime_env/plugin_schema_manager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SongGuyang) [reply python/ray/_private/runtime_env/plugin_schema_manager.py:65]: I have used `.json` instead of `schema.json`. Maybe we don't need this warning log?

**Comment by SongGuyang** [comment]
@architkulkarni take a look again？

**Review by architkulkarni — APPROVED** [review]
LGTM!

**Comment by architkulkarni** [comment]
Possibly relevant failure on Windows, `https://buildkite.com/ray-project/ray-builders-pr/builds/36944#0181af6f-c098-499c-adaf-8bb8bd465eed/6-9566` @SongGuyang do you mind taking a look?

**Comment by architkulkarni** [comment]
@mattip does anything obvious jump out at you for the Windows failure?

**Comment by stephanie-wang** [comment]
FYI, linux://python/ray/serve:test_runtime_env became significantly flakier after this merge.

Can you all review and decide whether we need to revert? @raulchen @SongGuyang @architkulkarni

**Comment by SongGuyang** [comment]
@stephanie-wang sorry about this! Let me see.

**Comment by SongGuyang** [comment]
I don't know the root cause in a short time. We can revert first.

**Pull request merged**